### PR TITLE
Obs encoding and vmap heuristics

### PR DESCRIPTION
**Pull request by moratodpg**
The main modifications include:

+ Agent id info removed from global_state
+ Agent id implemented as binary (default) and sinusoidal encoding 
+ Code for evaluating heuristics via vmap
+ Code for optimizing heuristics via vmap (defined in comb. chunks)

### CODE REVIEW

**Comment by prateekbhustali** [comment]
Thanks and great work @moratodpg! Results look good!

MAPPO (2 seeds) with same hyperparameters and various agent ID encodings

1. One-hot encoding : **-237.91**, -257.59 (imp-act/jaxMARL-mappo/gcwaf7d3)
2. Binary encoding: **-234.07**, -251.46 (imp-act/jaxMARL-mappo/mph8g9ku)
3. Sinusoidal encoding:  **-243.86**, -254.38 (imp-act/jaxMARL-mappo/4c24yqci)

### Binary Encoding

mappo_rnn_road_env_binary

### Sinusoidal Encoding

mappo_rnn_road_env_sinusoidal

**Comment by prateekbhustali** [comment]
## Heuristics

### ToyExample-v2

Evaluation completed in 16.05 seconds. (RTX 4090)
Best combo is (interval=6, threshold=1) with mean reward = -289.971863 and std = 159.434830

Top 5 combos (interval, threshold) by mean reward:
Rank 1: interval=6, threshold=1, mean_reward=-289.971863, done_any=True
Rank 2: interval=41, threshold=1, mean_reward=-290.443054, done_any=True
Rank 3: interval=43, threshold=1, mean_reward=-290.873383, done_any=True
Rank 4: interval=24, threshold=1, mean_reward=-291.835754, done_any=True
Rank 5: interval=42, threshold=1, mean_reward=-291.989532, done_any=True


### Cologne-v1

Evaluation completed in 210.31 seconds. (RTX 4090)
Best combo is (interval=42, threshold=1) with mean reward = -8649.166016 and std = 1864.961914

Top 5 combos (interval, threshold) by mean reward:
Rank 1: interval=42, threshold=1, mean_reward=-8649.166016, done_any=True
Rank 2: interval=24, threshold=1, mean_reward=-8656.776367, done_any=True
Rank 3: interval=30, threshold=1, mean_reward=-8676.398438, done_any=True
Rank 4: interval=50, threshold=1, mean_reward=-8676.826172, done_any=True
Rank 5: interval=21, threshold=1, mean_reward=-8682.662109, done_any=True

### CologneBonnDusseldorf-v1

(Needs ~2h, will run it later)

**Review by DanielHettegger — APPROVED** [review]
Looks great, thanks a lot Pablo! 🚀

**Comment by DanielHettegger** [comment]
I have added a few small tweaks so we can change the configs using hydra and the logs will be saved in the hydra output. 
For the ID encodings, I added the option to have no encoding and offset the binary IDs by 1 so it does not start at 0.

I think we can merge. Thanks again Pablo, very nice!

File: jaxmarl/environments/roadenv_wrapper.py [hunk 1]
```diff
@@ -49,6 +49,8 @@ def __init__(self, map_name, encoding_type="binary") -> None:
             self.agent_encodings = self.generate_binary_agent_ids(self.num_agents)
         elif encoding_type == "one-hot":
             self.agent_encodings = jnp.eye(self.num_agents, dtype=jnp.float32)
+        elif encoding_type == "none":
+            self.agent_encodings = jnp.zeros((self.num_agents, 0), dtype=jnp.float32)
```
Comment (prateekbhustali) [inline jaxmarl/environments/roadenv_wrapper.py:53]: If there is no encoding, can we skip it with a conditional? This will avoid big NNs. I am assuming this is for the case "what if we use the same policy for all agents?"

```
elif encoding_type == "none":
  self.agent_encodings = None

if self.agent_encodings is None:
 return jnp.concatenate([state.belief, _timestep, _budget], axis=1)
else:
 return jnp.concatenate([state.belief, _timestep, _budget, self.agent_encodings], axis=1)
```

Comment (moratodpg) [reply jaxmarl/environments/roadenv_wrapper.py:53]: Hey @omniscientoctopus, I think the current approach works (see attached). Note that we have a zero assigned and therefore the encoding dim. will remain 0. Later in the code, we call again agent_encodings `jnp.concatenate([state.belief, _timestep, _budget, self.agent_encodings], axis=1)` so I would go for the current version, which is cleaner. Would you agree as well or am I missing sth?
image

Comment (prateekbhustali) [reply jaxmarl/environments/roadenv_wrapper.py:53]: > Note that we have a zero assigned and therefore the encoding dim. will remain 0.

oh yeah, I missed that! I found the conditonal approach easier to understand, but I am ok with both 👍🏽

**Comment by DanielHettegger** [comment]
Ok, looks good. Then I will merge the branches now :)
